libraries/tools/kotlin-gradle-plugin/src/test/kotlin/org/jetbrains/kotlin/gradle/incremental/ClasspathSnapshotTestCommon.kt

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.gradle.incremental
+
+import com.google.gson.GsonBuilder
+import org.jetbrains.kotlin.test.MockLibraryUtil.compileKotlin
+import org.junit.Rule
+import org.junit.rules.TemporaryFolder
+import java.io.File
+
+abstract class ClasspathSnapshotTestCommon {
+
+    val testDataDir = File("libraries/tools/kotlin-gradle-plugin/src/testData/kotlin.incremental.useClasspathSnapshot/src")
+
+    @get:Rule
+    val tmp = TemporaryFolder()
+
+    abstract class RelativeFile(val baseDir: File, val relativePath: String) {
+        fun asFile() = File(baseDir, relativePath)
+    }
+
+    class SourceFile(baseDir: File, relativePath: String) : RelativeFile(baseDir, relativePath) {
+        init {
+            check(relativePath.endsWith(".kt") || relativePath.endsWith(".java"))
+        }
+    }
+
+    class ClassFile(baseDir: File, relativePath: String) : RelativeFile(baseDir, relativePath) {
+        init {
+            check(relativePath.endsWith(".class"))
+        }
+    }
+
+    protected fun SourceFile.replace(oldValue: String, newValue: String, valueMustExist: Boolean = true): SourceFile {
+        val fileContents = asFile().readText()
+        if (valueMustExist) {
+            check(fileContents.contains(oldValue)) { "String '$oldValue' not found in file '${asFile().path}'" }
+        }
+
+        val newSourceFile = SourceFile(tmp.newFolder(), relativePath).also { it.asFile().parentFile.mkdirs() }
+        newSourceFile.asFile().writeText(fileContents.replace(oldValue, newValue))
+        return newSourceFile
+    }
+
+    protected fun SourceFile.compile(): ClassFile {
+        val classesDir = tmp.newFolder()
+        when {
+            relativePath.endsWith(".kt") -> compileKotlin(asFile().path, classesDir)
```

Comment:
@nav-nav @gavra0 The current PR's tests are failing at this code path. The reason is that this code uses the existing `org.jetbrains.kotlin.test.MockLibraryUtil#compileKotlin()` method to compile a Kotlin class. However, that method at runtime will require `<kotlin-repo>/dist/kotlinc/lib/kotlin-compiler.jar` to be present, and that requires running `./gradlew dist` first before running the tests.

I also tried to use `K2JVMCompiler` directly with the following code:
```
    private fun compileKotlin(sourceFile: File, classesDir: File) {
        val args = listOf(sourceFile.path, "-d", classesDir.path)
        val err = ByteArrayOutputStream()
        K2JVMCompiler().exec(PrintStream(err), *args.toTypedArray())
        err.toString().let {
            if (it.isNotEmpty()) {
                error(it)
            }
        }
    }
```
but it fails with `NoClassDefFoundError: com/intellij/core/JavaCoreApplicationEnvironment`

If you have ideas how to fix the above, please let me know. Otherwise, I plan to add the .class files to the test data for now.

Reply:
I have updated the PR with .class files in test data. This PR has now passed TeamCity tests. Nataliya, could you help trigger an RRA run? Thanks! @nav-nav